### PR TITLE
Apply CircuitBreaker logic to PagingIterators.

### PR DESCRIPTION
**Pull request by matriv**
During a distributed operation the memory used in the PagingIterators
is not accounted at all which leads to OOMemoryExceptions which results
in killed nodes.

Added a new RamAccountingPageIterator which wraps the actual PagingIterator, calculates
the memory used and produces a CircuitBreakingException when the limits are exceeded.
So instead of having OOMemoryExecptions which kill node(s) of the cluster the job is killed
and the cluster is unaffected.

### CODE REVIEW

File: sql/src/main/java/io/crate/action/job/ContextPreparer.java [hunk 1]
```diff
@@ -502,35 +503,35 @@ public Boolean visitCountPhase(final CountPhase phase, final PreparerContext con
         }
 
         @Override
-        public Boolean visitMergePhase(final MergePhase phase, final PreparerContext context) {
+        public Boolean visitMergePhase(final MergePhase mergePhase, final PreparerContext context) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/action/job/ContextPreparer.java]: Unnecessary style change?

File: sql/src/main/java/io/crate/jobs/PageDownstreamContext.java [hunk 2]
```diff
@@ -155,7 +155,11 @@ private void triggerConsumer() {
     }
 
     private void mergeAndTriggerConsumer() {
-        mergeBuckets();
+        try {
+            mergeBuckets();
+        } catch (Throwable t) {
+            innerKill(t);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/jobs/PageDownstreamContext.java]: shouldn't this return here?

File: sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    final PagingIterator<Integer, Row> delegatePagingIterator;
+    private final RowAccounting rowAccounting;
+
+    private RamAccountingPageIterator(PagingIterator<Integer, Row> delegatePagingIterator,
+                                      RowAccounting rowAccounting) {
+        this.delegatePagingIterator = delegatePagingIterator;
+        this.rowAccounting = rowAccounting;
+    }
+
+    public static PagingIterator<Integer, Row> create(MergePhase mergePhase,
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java]: Would move this back into PagingIterators or maybe even into the interface itself. Having it here is missleading as the RamAccountingPageIterator isn't always used

Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java]: I'd change the interface to require a `Supplier<RowAccounting>` - this way you can remove the mergePhase again and depend on the concrete orderByPositions/numUpstreams. That should make testing a bit easier and the code is a bit less coupled to the mergePhase

File: sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                delegatePagingIterator,
+                new RowAccounting(
+                    mergePhase.inputTypes(),
+                    RamAccountingContext.forExecutionPhase(circuitBreaker, mergePhase)));
+        }
+        return delegatePagingIterator;
+    }
+
+    @VisibleForTesting
+    public static PagingIterator<Integer, Row> create(MergePhase mergePhase, RowAccounting rowAccounting) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java]: Please remove this. It's only used in testing and shouldn't be required here.

File: sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        return delegatePagingIterator.next();
+    }
+
+    @Override
+    public void remove() {
+        delegatePagingIterator.remove();
+    }
+
+    @Override
+    public void forEachRemaining(Consumer<? super Row> action) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java]: there are default implementations for forEachRemaining and remove - it shouldn't be necessary to implement them?

File: CHANGES.txt [hunk 6]
```diff
@@ -44,14 +44,19 @@ Changes
 Fixes
 =====
 
+ - Added proper handling when memory requirements of a distributed query
```
Comment (mfussenegger) [inline CHANGES.txt]: I'd rephrase this a bit. Only JOIN queries were really affected (or at least, that's the only place where this adds an improvement)

File: sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java (hunk 5 above)
Comment (matriv) [inline sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java]: But we need to forward the call to the inner delegate iterator, or am I missing something?

Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java]: The default implementation calls into `hasNext()` / `next()` and those calls will be forwarded to the delegate

File: CHANGES.txt [hunk 7]
```diff
@@ -44,10 +44,10 @@ Changes
 Fixes
 =====
 
- - Added proper handling when memory requirements of a distributed query
-   exceeds the available memory. Instead of OutOfMemoryException which leaded
-   to killed nodes in the cluster, the issue is detected and the query is
-   killed without affecting the cluster.
+ - Added proper handling when memory requirements of a ``JOIN`` query exceeds
+   the available memory. Instead of OutOfMemoryException which leaded to
```
Comment (mfussenegger) [inline CHANGES.txt]: "Instead of OutOfMemoryException" <- missing a verb?

File: sql/src/main/java/io/crate/operation/merge/PagingIterator.java [hunk 8]
```diff
@@ -45,4 +52,30 @@
...
+    /**
+     * Returns the suitable {@link PagingIterator} according to the use case.
+     * If requiresRepeat is true then the PagingIterator is wrapped with {@link RamAccountingPageIterator}
+     * which calculates the memory usage and applies CircuitBreaker logic.
+     */
+    static PagingIterator<Integer, Row> create(int numUpstreams,
+                                               boolean requiresRepeat,
+                                               @Nullable PositionalOrderBy orderBy,
+                                               Supplier<RowAccounting> rowAccountingSupplier) {
+        PagingIterator<Integer, Row> delegatePagingIterator;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/merge/PagingIterator.java]: minor: Would just call it `pagingIterator` - the `delegate` prefix doesn't add much readability here - if at all

File: sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java [hunk 9]
```diff
@@ -1,78 +1,35 @@
...
 
     @VisibleForTesting
     final PagingIterator<Integer, Row> delegatePagingIterator;
+    @Nullable
     private final RowAccounting rowAccounting;
 
-    private RamAccountingPageIterator(PagingIterator<Integer, Row> delegatePagingIterator,
-                                      RowAccounting rowAccounting) {
+    RamAccountingPageIterator(PagingIterator<Integer, Row> delegatePagingIterator,
+                                     @Nullable RowAccounting rowAccounting) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java]: why is this now nullable? This doesn't make sense to be used without a RowAccounting instance?

File: sql/src/test/java/io/crate/operation/merge/RamAccountingPageIteratorTest.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    public void testNoWrappingApplied() {
+        PagingIterator<Integer, Row> pagingIterator1 = PagingIterator.create(
+            2,
+            false,
+            null,
+            () -> null);
+
+        assertThat(pagingIterator1, instanceOf(PassThroughPagingIterator.class));
+
+        PositionalOrderBy positionalOrderBy = mock(PositionalOrderBy.class);
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/merge/RamAccountingPageIteratorTest.java]: why not create a real instance? It doesn't have any real dependencies...
